File: py-scouter/python/scouter/monitor/queueing_strategies/spc.py
from typing import Optional, Union

from scouter.integrations.http import HTTPConfig
from scouter.integrations.kafka import KafkaConfig
from scouter.integrations.rabbitmq import RabbitMQConfig
from scouter.monitor.queueing_strategies.base import BaseQueueingStrategy
from scouter.utils.logger import ScouterLogger

from ..._scouter import (  # pylint: disable=no-name-in-module
    Features,
    ServerRecords,
    SpcDriftProfile,
    SpcFeatureQueue,
)

logger = ScouterLogger.get_logger()


class SpcQueueingStrategy(BaseQueueingStrategy):
    def __init__(
        self,
        drift_profile: SpcDriftProfile,
        config: Union[KafkaConfig, HTTPConfig, RabbitMQConfig],
    ) -> None:
        """Initializes an SPC-specific queueing strategy with customized logic for inserting and
        publishing data related to statistical process control (SPC).

        This strategy manages the SPC drift profile and handles the queueing and publication
        of records for monitoring purposes.

        Args:
            drift_profile:
                Monitoring profile containing SPC feature drift profiles.
            config:
                Configuration for the monitoring producer. The configured producer
                will be used to publish drift records to the monitoring server.
        """
        super().__init__(config)
        self._feature_queue = SpcFeatureQueue(drift_profile=drift_profile)
        self._drift_profile = drift_profile

    def insert(self, features: Features) -> Optional[ServerRecords]:
        """Insert data into the monitoring queue.

        Args:
            features:
                List of features to insert into the monitoring queue.

        Returns:
            ServerRecords: The drift records published to the monitoring server.
        """
        try:
            self._feature_queue.insert(features)
            self._count += 1
            if self._count >= self._drift_profile.config.sample_size:
                return self._publish(self._feature_queue)
            return None
        except KeyError as exc:
            logger.error("Key error: {}", exc)
            return None

        except Exception as exc:  # pylint: disable=W0718
            logger.error("Failed to insert data into monitoring queue: {}. Passing", exc)
            return None

File: crates/scouter_client/src/queue/spc.rs
use pyo3::prelude::*;
use scouter_drift::spc::SpcFeatureQueue;
use scouter_error::ScouterError;
use scouter_events::producer::RustScouterProducer;
use scouter_types::spc::SpcDriftProfile;
use scouter_types::Features;
use tracing::{debug, error, instrument};

pub struct SpcQueue {
    queue: SpcFeatureQueue,
    producer: RustScouterProducer,
    count: usize,
}

impl SpcQueue {
    pub async fn new(
        drift_profile: SpcDriftProfile,
        config: &Bound<'_, PyAny>,
    ) -> Result<Self, ScouterError> {
        let producer = RustScouterProducer::new(config).await?;

        Ok(SpcQueue {
            queue: SpcFeatureQueue::new(drift_profile),
            producer,
            count: 0,
        })
    }

    #[instrument(skip_all, name = "Spc Insert", level = "debug")]
    pub async fn insert(&mut self, features: Features) -> Result<(), ScouterError> {
        let insert = self.queue.insert(features);

        // silently fail if insert fails
        if insert.is_err() {
            error!(
                "Failed to insert features into queue: {:?}",
                insert.unwrap_err().to_string()
            );
            return Ok(());
        }

        self.count += 1;

        debug!(
            "count: {}, sample_size: {}",
            self.count, self.queue.drift_profile.config.sample_size
        );

        if self.count >= self.queue.drift_profile.config.sample_size {
            let publish = self._publish().await;

            // silently fail if publish fails
            if publish.is_err() {
                // log error as string
                error!(
                    "Failed to publish drift records: {:?}",
                    publish.unwrap_err().to_string()
                );
                return Ok(());
            }

            self.count = 0;
        }

        Ok(())
    }

    async fn _publish(&mut self) -> Result<(), ScouterError> {
        let records = self.queue.create_drift_records()?;
        self.producer.publish(records).await?;
        self.queue.clear_queue();

        Ok(())
    }

    pub async fn flush(&mut self) -> Result<(), ScouterError> {
        self.producer.flush().await
    }
}
